src/Workspaces/Core/Portable/Storage/SQLite/v1/Interop/SqlConnection.cs

```diff
@@ -141,10 +138,32 @@ public ResettableSqlStatement GetResettableStatement(string query)
         {
             if (!_queryToStatement.TryGetValue(query, out var statement))
             {
-                var result = (Result)raw.sqlite3_prepare_v2(_handle, query, out var rawStatement);
+                using var _ = _handle.Lease();
+
+                var result = (Result)raw.sqlite3_prepare_v2(_handle.DangerousGetHandle(), query, out var rawStatement);
                 ThrowIfNotOk(result);
-                statement = new SqlStatement(this, rawStatement);
-                _queryToStatement[query] = statement;
+
+                SafeSqliteStatementHandle wrapper;
+                try
+                {
+                    wrapper = new SafeSqliteStatementHandle(_handle, rawStatement);
+                }
+                catch
+                {
+                    raw.sqlite3_finalize(rawStatement);
+                    throw;
+                }
+
+                try
+                {
+                    statement = new SqlStatement(this, wrapper);
+                    _queryToStatement[query] = statement;
```

Comment:
would it make sense to move this out to keep the try/catch minimal?

Reply:
This covers an OOM during dictionary resize.

Reply:
This has now been simplified.